cni/pkg/nodeagent/cni-watcher.go

```diff
@@ -153,26 +153,31 @@ func processAddEvent(body []byte) (CNIPluginAddEvent, error) {
 func (s *CniPluginServer) ReconcileCNIAddEvent(ctx context.Context, addCmd CNIPluginAddEvent) error {
 	log := log.WithLabels("cni-event", addCmd)
 
+	log.Debugf("netns: %s", addCmd.Netns)
+
 	// The CNI node plugin should have already checked the pod against the k8s API before forwarding us the event,
 	// but we have to invoke the K8S client anyway, so to be safe we check it again here to make sure we get the same result.
-	maxStaleRetries := 5
+	maxStaleRetries := 10
+	msInterval := 10
 	retries := 0
 	var ambientPod *corev1.Pod
 	var err error
+
+	log.Debugf("Checking pod: %s in ns: %s is enabled for ambient", addCmd.PodName, addCmd.PodNamespace)
 	// The plugin already consulted the k8s API - but on this end handler caches may be stale, so retry a few times if we get no pod.
-	for ambientPod, err = s.handlers.GetPodIfAmbient(addCmd.PodName, addCmd.PodNamespace); (ambientPod == nil) || (retries < maxStaleRetries); retries++ {
+	for ambientPod, err = s.handlers.GetPodIfAmbient(addCmd.PodName, addCmd.PodNamespace); (ambientPod == nil) && (retries < maxStaleRetries); retries++ {
```

Comment:
If a pod is not redirected by ambient, then GetPodIfAmbient always return nil. so  i understand it retries here. 

This is not expected

Reply:
Right. This was brought up and discussed a bit here: https://github.com/istio/istio/pull/48253#discussion_r1465364235

Originally this had no retry and was a simple check.

The difficulty is that we quite literally cannot tell the difference between "pod not redirected by ambient" and "informer cache is stale" - since the plugin does a direct query and the informer uses the caching machinery.

I'm happy to just skip the caching and use a raw kube client query here, like we do in the plugin:
```
func isAmbientPod(client kubernetes.Interface, podName, podNamespace string) (bool, error) {
	pod, err := client.CoreV1().Pods(podNamespace).Get(context.Background(), podName, metav1.GetOptions{})
	if err != nil {
		return false, err
	}
	ns, err := client.CoreV1().Namespaces().Get(context.Background(), podNamespace, metav1.GetOptions{})
	if err != nil {
		return false, err
	}

	return util.PodRedirectionEnabled(ns, pod), nil
}
```
I think that would be perfectly fine, thinking about this.

Reply:
Sending the entire pod spec with the CNI event would be easy, actually. The more I think about it the more I like that.

Reply:
I am not sure how can we pass the pod spec to cni plugin

Reply:
NVM, you mean send the pod spec from cni plugin to istio-cni server